functions/go/set-namespace/transformer/namespace.go

```diff
@@ -98,45 +71,152 @@ func (p *NamespaceTransformer) Config(o *fn.KubeObject) error {
 	return nil
 }
 
+// Transform replace the existing Namespace object, namespace-scoped resources' `metadata.name` and
+// other namespace reference fields to the new value.
 func (p *NamespaceTransformer) Transform(objects []*fn.KubeObject) {
 	p.SetDependsOnMap(objects)
 	for _, o := range objects {
 		switch {
+		// Skip local resource which `kpt live apply` skips.
+		case o.IsLocalConfig():
+			continue
 		case o.IsGVK("v1", "Namespace"):
-			o.SetName(p.Namespace)
+			o.SetName(p.NewNamespace)
 		case o.IsGVK("apiextensions.k8s.io/v1", "CustomResourceDefinition"):
-			err := o.Set(p.Namespace, "spec", "conversion", "webhook", "clientConfig", "service", "namespace")
-			if err != nil {
-				p.Errors = append(p.Errors, err.Error())
-			}
+			p.updateNamespace(o, "spec", "conversion", "webhook", "clientConfig", "service", "namespace")
 		case o.IsGVK("apiregistration.k8s.io/v1", "APIService"):
-			err := o.Set(p.Namespace, "spec", "service", "namespace")
-			if err != nil {
-				p.Errors = append(p.Errors, err.Error())
-			}
+			p.updateNamespace(o, "spec", "service", "namespace")
 		case o.GetKind() == "ClusterRoleBinding" || o.GetKind() == "RoleBinding":
 			var subjects []*Subject
 			o.GetOrDie(&subjects, "subjects")
 			for _, s := range subjects {
-				if s.Name == "default" {
-					s.Namespace = p.Namespace
+				if s.Kind == "ServiceAccount" {
+					if p.isSelectorMode() && s.Namespace != p.MatchingNamespace {
+						continue
+					}
+					s.Namespace = p.NewNamespace
 				}
 			}
 			o.SetOrDie(&subjects, "subjects")
 		case o.HasNamespace():
 			// Only update the namespace scoped resource. To determine if a resource is namespace scoped,
 			// we assume its namespace should have been set.
-			o.SetNamespace(p.Namespace)
+			p.updateNamespace(o, "metadata", "namespace")
 		default:
 			// skip the cluster scoped resource. We determine if a resource is cluster scoped by
 			// checking if the metadata.namespace is configured.
 		}
-		// Shall we accept custom fieldspec?
-		// p.SetCustomFieldSpec(o)
 		p.UpdateAnnotation(o)
 	}
 }
 
+// updateNamespace updates the namespace in three modes:
+// - Restrict mode: always update namespace from the given `fieldpaths`
+// - NsObjectSelector mode: update the namespace which matches the Namespace object found in resourceList.items.
+// - CustomSelector mode: update the namespace which matches the `namespaceSelector` in resourceList.functionConfig.
```

Comment:
Are NsObjectSelector and CustomSelector different for `updateNamespace`?  I feel like the logic could just be "if MatchingNamespace != ""`, then the old namespace must match.

Reply:
Thanks! So in `transform` we always update the namespace object. I like the simplification